File: src/core/pve_round_manager.py
from typing import *
from board import Board
from constant_types import UnitType
from player import Player
from utils import get_units_from_config, place_units_from_config, setup_board_from_config
from units import Unit


class PvERoundManager:
    """Manage progressive PvE rounds with different enemy configurations.

    Each configuration is a dict with keys `board_size` and `units` where
    `units` is a mapping of position -> `UnitType`. For backward compatibility
    the manager will also accept the legacy tuple form `(board_size, units)`.
    """

    def __init__(self, configs: List[Dict[str, Any]], initial_budget: int = 0):
        # store provided configs (dicts). Legacy tuple formats are accepted too.
        self.configs: List[Any] = configs

        # store player's initial units snapshot for resets
        # self.initial_player: Optional[List[Unit]] = None
        # optional stored starting positions for player and enemy units (list of (x,y))
        # self.player_positions: Optional[List[Tuple[int, int]]] = None
        # self.enemy_positions: Optional[List[Tuple[int, int]]] = None

        # Budget handling: initial and current player budget
        self.initial_budget = int(initial_budget) + self.configs[0]['budget_inc']
        self.player_budget = int(self.initial_budget)

        self.round_index = 0

    def _clone_unit_list(self, units: Optional[List[Unit]]) -> List[Unit]:
        if not units:
            return []
        out = []
        for u in units:
            c = u.clone()
            # ensure position cleared
            c.position = None
            out.append(c)
        return out

    def _add_config(self, units: List[Unit]):
        self.configs.append(self._clone_unit_list(units))

    def current_round(self) -> int:
        return self.round_index

    def num_rounds(self) -> int:
        return len(self.configs)

    def advance_round(self) -> bool:
        """Advance to the next configuration. Returns True if advanced."""
        if self.round_index + 1 < len(self.configs):
            self.round_index += 1
            # apply budget increment for the new round if specified
            cfg = self.configs[self.round_index]
            inc = 0
            if isinstance(cfg, dict):
                inc = int(cfg.get('budget_inc', 0) or 0)

            self.player_budget = int(self.player_budget) + int(inc)

            return True
        return False

    def reset_to_start(self):
        self.round_index = 0
        # reset budget to initial
        self.player_budget = int(self.initial_budget)


    # def save_player_positions(self, positions: List[Tuple[int, int]]):
    #     """Store the player's preferred starting positions (ordered list)."""
    #     if positions is None:
    #         self.player_positions = None
    #     else:
    #         self.player_positions = list(positions)

    # def save_enemy_positions(self, positions: List[Tuple[int, int]]):
    #     """Store the enemy preferred starting positions (ordered list)."""
    #     if positions is None:
    #         self.enemy_positions = None
    #     else:
    #         self.enemy_positions = list(positions)

    # def get_player_snapshot(self) -> List[Unit]:
    #     """Return fresh clones of the player's initial units."""
    #     return self._clone_unit_list(self.initial_player)
    
    def setup_round(self) -> Tuple[Board, List[Unit], List[Unit]]:
        """Setup round on a new board and return (board, enemy_units, player_units).
        """
        return setup_board_from_config(self.configs[self.round_index])

    def apply_round_to_board(self, board: Board, player1: Player, player2: Player) -> Tuple[Player, Player]:
        """Clear board and place enemy and player units for the current round.

        If `player` is provided, the player's units are placed on the board.
        If `player` is None, the stored initial player units are placed.

        Returns (enemy_units_list, player_units_list) that were placed on board.
        """
        # TODO: needs to behave differently on win or lose. 
        # On win, player units should preserve any permanent stacks (not implemented yet) gained from the previous round,
        # but on lose player should be reset to the stored snapshot. 
        board.reset_board()

        # prepare enemy units (fresh clones)
        enemy_units = place_units_from_config(board, self.configs[self.round_index], team=1)
        # set player's units_on_board from the returned list
        player1.set_units_on_board_from_list(enemy_units)

        # reset/place player2 units on board using dict API
        for u in player2.get_units_on_board_list():
            u.round_reset()
            board.place_board_unit(u, u.position)

                

        return player1, player2

File: csharp/MicroAutoChess.Core/Core/PvERoundManager.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection.Metadata;

namespace MicroAutoChess.Core
{
    public class PvERoundManager
    {
        // Configs may be LevelConfig objects or saved unit lists (List<Unit>)
        private List<object> _configs;
        public IReadOnlyList<object> Configs => _configs.AsReadOnly();

        public int InitialBudget { get; private set; }
        public int PlayerBudget { get; private set; }

        private int _roundIndex = 0;

        public PvERoundManager(List<object> configs, int initialBudget = 0)
        {
            _configs = configs ?? new List<object>();
            int firstInc = 0;
            if (_configs.Count > 0 && _configs[0] is LevelConfig lc) firstInc = lc.BudgetInc;
            InitialBudget = initialBudget + firstInc;
            PlayerBudget = InitialBudget;
            _roundIndex = 0;
        }

        private List<Unit> CloneUnitList(List<Unit>? units)
        {
            if (units == null) return new List<Unit>();
            var outList = new List<Unit>();
            foreach (var u in units)
            {
                var c = u.Clone();
                c.Position = null;
                outList.Add(c);
            }
            return outList;
        }

        public void AddConfig(List<Unit> units)
        {
            _configs.Add(CloneUnitList(units));
        }

        public int CurrentRound() => _roundIndex;
        public int NumRounds() => _configs.Count;

        public bool AdvanceRound()
        {
            if (_roundIndex + 1 < _configs.Count)
            {
                _roundIndex += 1;
                int inc = 0;
                if (_configs[_roundIndex] is LevelConfig lc) inc = lc.BudgetInc;
                PlayerBudget = PlayerBudget + inc;
                return true;
            }
            return false;
        }

        public void ResetToStart()
        {
            _roundIndex = 0;
            PlayerBudget = InitialBudget;
        }

        // Setup round returns (board, enemyUnits, playerUnits)
        public (Board, List<Unit>, List<Unit>) SetupRound()
        {
            if (_configs.Count == 0) throw new InvalidOperationException("No configs available");
            var cfg = _configs[_roundIndex];
            if (cfg is LevelConfig lc) return SetupBoardFromConfig(lc);
            if (cfg is List<Unit> units) // treat as snapshot list
            {
                var board = new HexBoard((7,8));
                var placed = new List<Unit>();
                foreach (var u in units)
                {
                    var nu = u.Clone();
                    nu.Position = u.Position;
                    if (nu.Position != null && board.IsValidPosition(nu.Position.Value))
                    {
                        board.PlaceBoardUnit(nu, nu.Position.Value);
                        placed.Add(nu);
                    }
                }
                return (board, placed, new List<Unit>());
            }
            throw new ArgumentException("Unsupported config type");
        }

        public (Board, Player, Player) ApplyRoundToBoard(Board board, Player player1, Player player2)
        {
            // clear board
            board.ResetBoard();

            // prepare enemy units (fresh clones) and place them as team 1
            var cfg = _configs[_roundIndex];
            List<Unit> enemyUnits = new List<Unit>();
            if (cfg is LevelConfig lc)
            {
                enemyUnits = PlaceUnitsFromConfig(board, lc, team: Team.TEAM_1);
            }
            else if (cfg is List<Unit> unitList)
            {
                // re-place provided units (assume they are pre-positioned)
                foreach (var u in unitList)
                {
                    var nu = u.Clone();
                    nu.Team = Team.TEAM_1;
                    if (nu.Position != null)
                    {
                        board.PlaceBoardUnit(nu, nu.Position.Value);
                        enemyUnits.Add(nu);
                    }
                }
            }

            // set player1 units-on-board from enemyUnits (mirror Python behavior where team1 is enemy)
            player1.UnitsOnBoard.Clear();
            foreach (var eu in enemyUnits)
            {
                if (eu.Position != null) player1.UnitsOnBoard[eu.Position.Value] = eu;
            }

            // reset/place player2 units on board using player's UnitsOnBoard dictionary
            foreach (var kv in player2.UnitsOnBoard)
            {
                var u = kv.Value;
                if (u == null) continue;
                u.RoundReset();
                if (u.Position != null) board.PlaceBoardUnit(u, u.Position.Value);
            }

            return (board, player1, player2);
        }

        // Helper: create board and place units from LevelConfig
        private (Board, List<Unit>, List<Unit>) SetupBoardFromConfig(LevelConfig cfg)
        {
            Board board;
            if (cfg.BoardType == "hex") board = new HexBoard(cfg.BoardSize);
            else if (cfg.BoardType == "square") board = new SquareBoard(cfg.BoardSize);
            else if (cfg.BoardType == "diagonal_square") board = new DiagonalSquareBoard(cfg.BoardSize);
            else board = new HexBoard(cfg.BoardSize);

            var enemy = PlaceUnitsFromConfig(board, cfg, team: Team.TEAM_1);
            return (board, enemy, new List<Unit>());
        }

        private List<Unit> PlaceUnitsFromConfig(Board board, LevelConfig cfg, Team team = Team.TEAM_1)
        {
            var placed = new List<Unit>();
            foreach (var kv in cfg.Units)
            {
                var pos = kv.Key;
                var ut = kv.Value;
                if (board.IsValidPosition(pos) && board.IsEmpty(pos))
                {
                    var u = new Unit(ut, UnitRarity.COMMON, team) { Position = pos };
                    u.CurrentHealth = u.GetMaxHealth();
                    board.PlaceBoardUnit(u, pos);
                    placed.Add(u);
                }
            }
            return placed;
        }

        // Spend budget; returns true if there was enough budget
        public bool SpendBudget(int amount)
        {
            if (amount <= 0) return true;
            if (PlayerBudget >= amount)
            {
                PlayerBudget -= amount;
                return true;
            }
            return false;
        }

        // Add budget (refunds, rewards)
        public void AddBudget(int amount)
        {
            if (amount <= 0) return;
            PlayerBudget += amount;
        }
    }
}
